sdk/sync/types.go

```diff
@@ -1,43 +1,106 @@
 package sync
 
 import (
+	"context"
+	"encoding/json"
 	"fmt"
 	"reflect"
-	"strings"
+
+	"github.com/ipfs/testground/sdk/runtime"
 )
 
-// A Subtree represents a subtree of the sync tree for a test run. It is bound
-// to a particular subpath pattern, where payloads of a specific type are posted
-// and manipulated.
-type Subtree struct {
-	// GroupKey is the name of the group the keys in this subtree belong to.
-	GroupKey string
+// State represents a state in a distributed state machine, identified by a
+// unique string within the test case.
+type State string
 
-	// PayloadType is the type of the payload. Must be a pointer type.
-	PayloadType reflect.Type
+// Key gets the Redis key for this State, contextualized to a set of RunParams.
+func (s State) Key(rp *runtime.RunParams) string {
+	p := fmt.Sprintf("run:%s:plan:%s:case:%s:states:%s", rp.TestRun, rp.TestPlan, rp.TestCase, string(s))
+	return p
+}
 
-	// KeyFunc returns the key of this entry within the subtree, optionally
-	// deriving it from the supplied payload.
-	KeyFunc func(payload interface{}) string
+// Barrier represents a barrier over a State. A Barrier is a synchronisation
+// checkpoint that will fire once the `target` number of entries on that state
+// have been registered.
+type Barrier struct {
+	C chan error
+
+	ctx    context.Context
+	state  State
+	key    string
+	target int64
 }
 
-// State represents a state in a distributed test.
-type State string
+// Topic represents a meeting place for test instances to exchange arbitrary
+// data.
+type Topic struct {
+	name string
+	typ  reflect.Type
+}
+
+// NewTopic constructs a Topic with the provided name, and the type of the
+// supplied value, derived via reflect.TypeOf, unless the supplied value is
+// already a reflect.Type. This method does not retain actual value from which
+// the type is derived.
+func NewTopic(name string, typ interface{}) *Topic {
+	t, ok := typ.(reflect.Type)
+	if !ok {
+		t = reflect.TypeOf(typ)
+	}
+	return &Topic{name, t}
+}
```

Comment:
@Stebalien have a look at the topic constructor, and the downstream simplifications. Good suggestion!